### PR TITLE
Reactive Updates Refactor

### PR DESCRIPTION
**Pull request by samuelselleck**
Quick overview of changes:
The render tree is now saved between frames and children are only re-created if their parent component finds it necessary based on property changes (paves the way towards reactive updates instead of an update tick). Slot has been stabilized and arbitrary hierarchies of repeat and stacker elements are working for all examples.

### CODE REVIEW

File: pax-core/src/engine.rs [hunk 1]
```diff
@@ -643,69 +263,59 @@ impl PaxEngine {
     /// 3. Render:
     ///     a. find lowest node (last child of last node)
     ///     b. start rendering, from lowest node on-up, throughout tree
-    pub fn tick(&self, rcs: &mut HashMap<String, Box<dyn RenderContext>>) -> Vec<NativeMessage> {
-        let root_component_instance = Rc::clone(&self.main_component);
-        let mut z_index = LayerId::new(None);
+    pub fn tick(&mut self) -> Vec<NativeMessage> {
```
Comment (zackbrown) [inline pax-core/src/engine.rs:266]: Please propagate this refactor to `pax-chassis-common` and get it building again

File: pax-core/src/rendering.rs [hunk 2]
```diff
@@ -209,22 +208,34 @@ pub trait InstanceNode {
...
 
-    /// Expands the current `InstanceNode` into a stateful `ExpandedNode`, with its own instances of properties & common properties, in the context of the
-    /// provided `PropertiesTreeContext`.  Node expansion takes into account the "parallel selves" that an `InstanceNode` may have through the
-    /// lens of declarative control flow, [`ConditionalInstance`] and [`RepeatInstance`].
+    // /// Used by elements that need to communicate across native rendering bridge (for example: Text, Clipping masks, scroll containers)
+    // /// Called by engine after [`expand_node`], passed calculated size and transform matrix coefficients for convenience
+    // /// Expected to induce side-effects (if appropriate) via enqueueing messages to the native message queue
+    // ///
+    // /// An implementor of `handle_native_patches` is responsible for determining which properties if any have changed
+    // /// (e.g. by keeping a local patch object as a cache of last known values.)
```
Comment (zackbrown) [inline pax-core/src/rendering.rs:216]: Can these be plain `///` rustdoc comments, or were the `//`s added to hack around a cargo error?

File: pax-core/src/rendering.rs [hunk 3]
```diff
@@ -244,8 +256,12 @@ pub trait InstanceNode {
     /// Occurs in a post-order traversal of the render tree. Most primitives
     /// are expected to draw their contents to the rendering context during this event.
     #[allow(unused_variables)]
-    fn handle_render(&self, rtc: &mut RenderTreeContext, rc: &mut Box<dyn RenderContext>) {
-        //no-op default implementation
+    fn render(
```
Comment (zackbrown) [inline pax-core/src/rendering.rs:259]: can we name `render` and `draw` consistently?  (one or the other)

File: pax-core/src/engine/expanded_node.rs [hunk 4]
```diff
@@ -0,0 +1,619 @@
...
+            self.instance_template.handle_unmount(&self, context);
+        }
+    }
+
+    /// Manages unpacking an Rc<RefCell<dyn Any>>, downcasting into
+    /// the parameterized `target_type`, and executing a provided closure `body` in the
+    /// context of that unwrapped variant (including support for mutable operations),
+    /// the closure is executed.  Used at least by calculating properties in `expand_node` and
+    /// passing `&mut self` into event handlers (where the typed `self` is retrieved from an instance of `dyn Any`)
+    pub fn with_properties_unwrapped<T: 'static, R>(
```
Comment (zackbrown) [inline pax-core/src/engine/expanded_node.rs:334]: nice refactor away from macro! ✨

File: pax-core/src/slot.rs [hunk 5]
```diff
@@ -42,65 +42,59 @@ impl InstanceNode for SlotInstance {
...
             }
-        }
-
-        this_expanded_node
+        });
     }
 
+    // fn handle_unmount(&self, expanded_node: &Rc<ExpandedNode>, context: &mut RuntimeContext) {
+    //     expanded_node.set_children(vec![], context);
+    // }
```
Comment (zackbrown) [inline pax-core/src/slot.rs:96]: can delete?

File: pax-core/src/declarative_macros.rs [hunk 6]
```diff
@@ -1,50 +1,29 @@
...
-        let current_prop = &mut *$var.$field.as_mut();
-        if let Some(vtable_id) = current_prop._get_vtable_id() {
-            let new_value_wrapped: Box<dyn Any> =
-                $ptc.compute_vtable_value(&$expanded_node, vtable_id);
-            if let Ok(downcast_value) = new_value_wrapped.downcast::<$inner_type>() {
-                current_prop.set(*downcast_value);
-            } else {
-                panic!()
-            } //downcast failed
+pub fn handle_vtable_update<V: Default + Clone + std::fmt::Debug + 'static>(
```
Comment (zackbrown) [inline pax-core/src/declarative_macros.rs:13]: nice refactor! 🌈

File: pax-std/pax-std-primitives/src/lib.rs [hunk 7]
```diff
@@ -2,7 +2,7 @@
 pub mod ellipse;
 pub mod frame;
 pub mod group;
-// pub mod image;
+pub mod image;
```
Comment (zackbrown) [inline pax-std/pax-std-primitives/src/lib.rs:5]: Thank you!

**Review by zackbrown — CHANGES_REQUESTED** [review]
Excellent work!  Requesting a handful of minor changes.  Overall great design and very welcome cleaning & refactoring.

File: examples/src/component_in_component/Cargo.toml [hunk 8]
```diff
@@ -6,8 +6,9 @@ default-run = "run"
 
 [dependencies]
 pax-lang = { version = "0.10.10" , path = ".pax/pkg/pax-lang" }
-pax-std = { version = "0.10.10" , path = ".pax/pkg/pax-std" }
-pax-compiler = { version = "0.10.10", optional = true, path = ".pax/pkg/pax-compiler" }
+pax-std = { version = "0.10.10"  , path = ".pax/pkg/pax-std" }
```
[the remark this answers is not in the document]
Comment (samuelselleck) [reply examples/src/component_in_component/Cargo.toml]: Okay! Won't commit these in the future :)

File: pax-core/src/engine/expanded_node.rs (hunk 4 above)
Comment (samuelselleck) [reply pax-core/src/engine/expanded_node.rs:340]: <3

File: pax-core/src/slot.rs (hunk 5 above)
Comment (samuelselleck) [reply pax-core/src/slot.rs]: 🧹

File: pax-std/pax-std-primitives/src/lib.rs (hunk 7 above)
Comment (samuelselleck) [reply pax-std/pax-std-primitives/src/lib.rs:5]: <3

**Comment by samuelselleck** [comment]
Thanks for the feedback Zack!

File: pax-core/src/rendering.rs [hunk 9]
```diff
@@ -346,25 +354,11 @@ impl BaseInstance {
...
-            &(self.instance_prototypical_common_properties_factory)(),
-        )
-    }
-
     /// Return the list of instance nodes that are children of this one.  Intuitively, this will return
     /// instance nodes mapping exactly to the template node definitions.
     /// For `Component`s, `get_instance_children` returns the root(s) of its template, not its `slot_children`.
     /// (see [`get_slot_children`] for the way to retrieve the latter.)
-    pub fn get_children(&self) -> &InstanceNodePtrList {
+    pub fn get_template_children(&self) -> &InstanceNodePtrList {
```
[the remark this answers is not in the document]
Comment (samuelselleck) [reply pax-core/src/rendering.rs:361]: Noted! I remember you talking about this before but forgot about it.

File: pax-core/src/engine/expanded_node.rs [hunk 10]
```diff
@@ -0,0 +1,619 @@
...
+        dispatch_context_menu,
+        ArgsContextMenu,
+        context_menu_handlers
+    );
+    dispatch_event_handler!(dispatch_click, ArgsClick, click_handlers);
+    dispatch_event_handler!(dispatch_wheel, ArgsWheel, wheel_handlers);
+}
+
+#[cfg_attr(debug_assertions, derive(Debug))]
+pub struct ComputedExpandedProperties {
```
[the remark this answers is not in the document]
Comment (samuelselleck) [reply pax-core/src/engine/expanded_node.rs:512]: I agree, LayoutProperties is a good name.

File: pax-core/src/rendering.rs (hunk 2 above)
Comment (samuelselleck) [reply pax-core/src/rendering.rs:216]: Oops! Not intended.

File: pax-core/src/engine.rs (hunk 1 above)
Comment (samuelselleck) [reply pax-core/src/engine.rs:266]: Got it building! Isn't rendering text, is this expected?

Comment (zackbrown) [reply pax-core/src/engine.rs:266]: Previously it was rendering text (see screenshot from mid-Dec) but not a critical issue to fix rn.  Probably related to changes in native message ordering, or possibly occlusion (which is yet-unimplemented on macos/ios).  We can fix this later.

Pasted Graphic 1

File: pax-core/src/rendering.rs (hunk 3 above)
Comment (zackbrown) [reply pax-core/src/rendering.rs:245]: This approach is fine for now and thanks for the rename!

**Pull request merged**